ts/nni_manager/common/trial_keeper/keeper.ts

```diff
@@ -0,0 +1,175 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+/**
+ *  The helper class to implement a "reuse" training service.
+ *
+ *  TrialKeeper has a very similar interface to TrainingSerivceV3.
+ *  In fact the local training service is a thin wrapper over TrialKeeper.
+ **/
+
+import { EventEmitter } from 'events';
+import fs from 'fs/promises';
+import path from 'path';
+
+import { HttpChannelServer } from 'common/command_channel/http';
+import globals from 'common/globals';
+import { GpuScheduler } from 'common/gpu_scheduler';
+import { Logger, getLogger } from 'common/log';
+import { TrialProcess } from './process';
+
+export declare namespace TrialKeeper {
+    export interface TrialOptions {
+        id: string;
+        command: string;
+        codeDirectoryName: string;
+        sequenceId?: number;
+        gpuNumber?: number;
+        gpuRestrictions?: GpuRestrictions;
+    }
+
+    export interface GpuRestrictions {
+        onlyUseIndices?: number[];
+        rejectActive?: boolean;
+        rejectComputeActive?: boolean;
+    }
+}
+
+export class TrialKeeper {
+    private envId: string;
+    private channels: HttpChannelServer;
+    private dirs: Map<string, string> = new Map();
+    private emitter: EventEmitter = new EventEmitter();
+    private gpuScheduler: GpuScheduler | null = null;
+    private log: Logger;
+    private platform: string;
+    private trials: Map<string, TrialProcess> = new Map();
+
+    constructor(environmentId: string, platform: string, enableGpuScheduler: boolean) {
+        this.envId = environmentId;
+        this.platform = platform;
+        this.log = getLogger(`TrialKeeper.${environmentId}`);
+
+        this.channels = new HttpChannelServer(this.envId, `/env/${this.envId}`);
+
+        if (enableGpuScheduler) {
+            this.gpuScheduler = new GpuScheduler();
+        }
+
+        this.channels.onReceive((trialId, command) => {
+            if (command.type === 'request_parameter') {
+                this.emitter.emit('request-parameter', trialId);
+            } else if (command.type === 'metric') {
+                this.emitter.emit('metric', trialId, command.metric);
+            } else {
+                this.log.error('Received bad command:', trialId, command);
+            }
+        });
+    }
+
+    public async start(): Promise<void> {
+        await this.channels.start();
+
+        if (this.gpuScheduler !== null) {
```

Comment:
How does the scheulder work for pure-CPU workloads?

Reply:
Always accept. This is the v2 behavior.